### PR TITLE
Add Unit Tests and Update Notice Generation for MQTT Connector's Azure Function

### LINKED ISSUE #25: [enhancement] Unit Tests for the MQTT Connector's Azure Function
## Description

Include unit tests for Azure Functions

## Acceptance criteria

- [x] Can be executed by workflows
- [x] Can run locally

### PR DESCRIPTION
**Pull request by jorchiu**
This PR closes #25

### CODE REVIEW

File: cloud_connectors/azure/mqtt_connector/res/azure_function/src/run.cs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+            List<Type> dataTypes = new List<Type>() { typeof(Double), typeof(Boolean), typeof(Int32) };
+            var jsonPatchDocument = new JsonPatchDocument();
+            string errorMessage = null;
+
+            foreach (Type type in dataTypes)
+            {
+                try
+                {
+                    // Parse the data to a type
+                    dynamic value = TypeDescriptor.GetConverter(type).ConvertFromInvariantString(instance.data);
```
Comment (ashbeitz) [inline cloud_connectors/azure/mqtt_connector/res/azure_function/src/run.cs:57]: "value" isn't a great identifier name here, as it seems to contain a type.

File: cloud_connectors/azure/mqtt_connector/res/azure_function/src/run.cs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                }
+                catch(RequestFailedException ex)
+                {
+                    errorMessage = @$"Failed to parse {instance.data} due to {ex.Message}.
+                        Cannot set instance {instance.instance_id}{instance.instance_property_path}
+                        based on model {instance.model_id} to {instance.data}";
+                }
+            }
+
+            throw new NotSupportedException(errorMessage);
```
Comment (ashbeitz) [inline cloud_connectors/azure/mqtt_connector/res/azure_function/src/run.cs:87]: I don't like the method ending with a throw.  It seems very pessimistic.  Can you please move this into the catch block?  Also, it seems like there is no need for errorMessage to be declared outside of this block.

File: cloud_connectors/azure/mqtt_connector/README.md [hunk 3]
```diff
@@ -136,9 +136,9 @@ Navigate to the client that you have created in your Event Grid Namespace, and v
 
 1. [Create an Azure Function app](https://learn.microsoft.com/en-us/azure/event-grid/custom-event-to-function#create-azure-function-app) that triggers your Azure Event Grid. Ensure you set the Runtime stack to .NET and version 6.0.
 
-1. Replace the code in your Azure Function `run.cs` with the code in the `res/azure_function/run.cs` folder of this repo.
+1. Replace the code in your Azure Function with the code in the `{freyja-root-dir}/cloud_connectors/azure/mqtt_connector/res/azure_function/src/run.cs` file.
 
-1. Add the following files `res/azure_function/function.json` and `res/azure_function/function.csproj` to your Azure Function.
+1. Add the following files `{freyja-root-dir}/cloud_connectors/azure/mqtt_connector/res/azure_function/res/src/function.json` and `{freyja-root-dir}/cloud_connectors/azure/mqtt_connector/res/azure_function/src/function.csproj` to your Azure Function.
```
Comment (wilyle) [inline cloud_connectors/azure/mqtt_connector/README.md:141]: Let's simplify this a bit:

Add `res/src/function.json` and `src/function.csproj` from `{freyja-root-dir}/{whatever}/azure_function` to your Azure Function

File: cloud_connectors/azure/mqtt_connector/res/azure_function/src/run.cs [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+                catch (Exception ex) when (ex is NotSupportedException || ex is ArgumentException || ex is FormatException)
+                {
+                    continue;
+                }
+
+                try
+                {
+                    // Exit the function if the instance is successfully updated.
+                    await client.UpdateDigitalTwinAsync(instance.instance_id, jsonPatchDocument);
+                    return;
```
Comment (ashbeitz) [inline cloud_connectors/azure/mqtt_connector/res/azure_function/src/run.cs:77]: If you fix the throw issue in my next comment, then there is no need to return here.

File: cloud_connectors/azure/mqtt_connector/res/azure_function/src/run.cs [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        /// An Azure Function that updates an Azure Digital Twin based on the request.
+        /// </summary>
+        /// <param name="cloudEvent">the cloudEvent request that is received.</param>
+        /// <param name="logger">the logger</param>
+        /// <exception>An exception is thrown if the Azure Digital Twin client cannot update an instance.</exception>
+        /// <returns></returns>
+        [FunctionName("MQTTConnectorAzureFunction")]
+        public static async Task Run([EventGridTrigger] CloudEvent cloudEvent, ILogger logger)
+        {
+            List<Type> dataTypes = new List<Type>() { typeof(Double), typeof(Boolean), typeof(Int32) };
```
Comment (wilyle) [inline cloud_connectors/azure/mqtt_connector/res/azure_function/src/run.cs:100]: I don't think this is being used

**Review by ashbeitz — COMMENTED** [review]
Please take a look at my comments.

File: cloud_connectors/azure/mqtt_connector/res/azure_function/tests/MQTTConnectorAzureFunction.Tests.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        {
+            _instance.data = "44.5";
+            await MQTTConnectorAzureFunction.UpdateDigitalTwinAsync(_client, _instance);
+            Assert.Pass();
+        }
+
+        [Test]
+        public void UpdateDigitalTwinAsync_ThrowNotSupported()
+        {
+            Assert.ThrowsAsync<NotSupportedException>(async () => await MQTTConnectorAzureFunction.UpdateDigitalTwinAsync(_client, _instance));
```
Comment (wilyle) [inline cloud_connectors/azure/mqtt_connector/res/azure_function/tests/MQTTConnectorAzureFunction.Tests.cs:41]: can we add a comment to say that this is testing when data == null? or just manually set it, even though that's already the default? it's a bit unclear to me what this is testing

File: tools/dotnet_notice_generation.sh [hunk 7]
```diff
@@ -1,36 +1,44 @@
...
+mkdir -p "$dotnet_licenses_output_directory"
 echo "Getting the .NET Third Party licenses"
-dotnet-project-licenses -i $DOTNET_DIRECTORY -o -f "$DOTNET_DIRECTORY/dotnet_licenses_output" -u --json -e -c \
---licenseurl-to-license-mappings "$DOTNET_DIRECTORY/license_url_to_type.json"
-./tools/dotnet_get_licenses.sh "$DOTNET_DIRECTORY/dotnet_licenses_output/licenses.json" "$DOTNET_DIRECTORY/dotnet_licenses_output"
-./tools/dotnet_append_to_notice.sh "$notice_file_path" "$DOTNET_DIRECTORY/dotnet_licenses_output/licenses.json"
-rm -r "$DOTNET_DIRECTORY/dotnet_licenses_output"
+
+dotnet-project-licenses -i $dotnet_directory -o -f "$dotnet_directory/dotnet_licenses_output" -u --json -e -c \
+--licenseurl-to-license-mappings "$license_url_to_license_mappings"
```
Comment (wilyle) [inline tools/dotnet_notice_generation.sh:37]: indent this line

File: cloud_connectors/azure/mqtt_connector/res/azure_function/src/run.cs (hunk 1 above)
Comment (jorchiu) [reply cloud_connectors/azure/mqtt_connector/res/azure_function/src/run.cs]: Fixed. Renamed to convertedDataType

File: cloud_connectors/azure/mqtt_connector/res/azure_function/tests/MQTTConnectorAzureFunction.Tests.cs (hunk 6 above)
Comment (jorchiu) [reply cloud_connectors/azure/mqtt_connector/res/azure_function/tests/MQTTConnectorAzureFunction.Tests.cs]: Fixed. I've manually set _instance.data = null here.

File: cloud_connectors/azure/mqtt_connector/res/azure_function/src/run.cs (hunk 4 above)
Comment (jorchiu) [reply cloud_connectors/azure/mqtt_connector/res/azure_function/src/run.cs]: I've removed the `return;` statement.

File: cloud_connectors/azure/mqtt_connector/res/azure_function/src/run.cs (hunk 2 above)
Comment (jorchiu) [reply cloud_connectors/azure/mqtt_connector/res/azure_function/src/run.cs]: Fixed. I ended up refactoring this function, and removed the loop that tries to do a data type conversion.

**Review by ashbeitz — APPROVED** [review]
Approved.

File: cloud_connectors/azure/mqtt_connector/README.md (hunk 3 above)
Comment (wilyle) [reply cloud_connectors/azure/mqtt_connector/README.md]: Please look at my suggestion again, the current wording is incorrect grammar

File: cloud_connectors/azure/mqtt_connector/res/azure_function/src/run.cs [hunk 8]
```diff
@@ -26,65 +26,81 @@ public class DigitalTwinsInstance {
...
-                // Try to parse the data using the next type
-                catch (Exception ex) when (ex is NotSupportedException || ex is ArgumentException || ex is FormatException)
-                {
-                    continue;
-                }
+            try {
+                // Get the converter for the instance's data by using the data type name
+                // then use that converter to change the instance's data type to the converter's type.
+                string dataTypeConverterName = ConvertStringToDataType(dataTypeName);
+                Type type = Type.GetType(dataTypeConverterName);
```
Comment (wilyle) [inline cloud_connectors/azure/mqtt_connector/res/azure_function/src/run.cs:81]: Why change the map to `Dictionary<string, string>`? If we leave it as `Dictionary<string, Type>` we don't need this

Comment (jorchiu) [reply cloud_connectors/azure/mqtt_connector/res/azure_function/src/run.cs]: Fixed. I've changed it to `Dictionary<string, type>

**Review by wilyle — APPROVED** [review]

**Comment by jorchiu** [comment]
This PR closes #16
